### ISSUE TITLE
The JWT option on VirtualService cannot anymore be disabled at the route level

### ISSUE THREAD

**Issue by bcollard** [solo-io/gloo#5917]
### Gloo Edge Version

1.10.x (latest stable)

### Describe the bug

Given a JWT verification configured at the VirtualService level
and a route with JWT option disabled
then the request matching this route is still subject to JWT validation
Also, the JWT option at the VS level is using `claimsToHeader`, so might be related to the filter flow when clear route cache is used.

Works as expected until Gloo EE version `1.10.0-beta8`
It doesn't work anymore since Gloo EE version `1.10.0-rc1`

### Steps to reproduce the bug

see the full test in this gist: https://gist.github.com/bcollard/75be986596bc9c6c9b066e14ae31d5bd

### Expected Behavior

no JWT verification on the route where JWT is disabled

**Comment by ben-taussig-solo** [comment]
Update: 
 - This issue appears to be the result of changes to the behavior of the `extensions.filters.http.jwt_authn.v3.JwtRequirement` filter in Envoy v1.20.0. Unfortunately, these changes are not, as far as we can tell, listed in the v1.20.0 changelog and we have yet to identify them.
   - Essentially, as of envoy v1.20.0/solo-projects v1.10.0-rc1, the envoy [`JwtRequirement`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto.html?highlight=allow_missing#extensions-filters-http-jwt-authn-v3-jwtrequirement) filter's [`allow_missing_or_failed`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto.html?highlight=allow_missing#envoy-v3-api-field-extensions-filters-http-jwt-authn-v3-jwtrequirement-allow-missing-or-failed) setting now must be enabled in order to allow JWT verification to be disabled on a specific route.
 -  For now, you may be able to work around the issue by setting `virtualHost.options.jwtStaged.beforeExtAuth.allowMissingOrFailedJwt` to true.

**Comment by kevin-shelaga** [comment]
The proposed workaround would open all routes to allow a missing or failed jwt, which is not ideal for my use case.

**Comment by ben-taussig-solo** [comment]
Some clarification -- the JWT filter can be disabled at route level when the route specifies a `routeAction`. Take, for example, the following virtual service configured with the other resources present in the gist linked in the original issue:
```
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: httpbin
  namespace: gloo-system
spec:
  virtualHost:
    domains:
    - '*'
    routes:
    - matchers: # the end-user didn't sent a JWT in the Authorization header, so we disable JWT validation (which is set at the VS level), and we redirect the user to some other page (here /status/401)
      - prefix: /
        headers:
        - name: authorization
          invertMatch: true
      options:
        jwtStaged:
          beforeExtAuth:
            disable: true
      routeAction:
        single:
          upstream:
            name: default-httpbin-8000
            namespace: gloo-system
    - matchers: # the default route, picked at first when the user comes with the Authorization header (containing a JWT)
      - prefix: /
      options:
        prefixRewrite: /ip # dummy route, not used because of the dynamic route selection after a claim is extracted from the JWT
      routeAction:
        single:
          upstream:
            name: default-httpbin-8000
            namespace: gloo-system
    options:
      jwtStaged:
        beforeExtAuth:
          providers:
            inline:
              jwks:
                local:
                  key: |
                    -----BEGIN PUBLIC KEY-----
                    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA6u+z4YAGfkMKX5EJJIk+
                    jcX91JBfl2iUVcrC1NOYza8K5rQbOb0eg8fy3GiITZ/e8V5wmWf90Fr8z+3rhy1x
                    ehcH+S3usNFfDE01rhGmofGOvUXAvzBvDgM+9s8zW1hi7HZnNi4eXagIYxirTD4Y
                    ANQq02IhUr9/IQbsNpwshPmDsav59ApDn69jw9WcWW6kiJerA+mvlw+xOzGfl2RA
                    OB8f7zO11yuSYrXgy7ap4jvrkV6FMkqf6fLpd1mOiNGEENixSMUjoXRPsZ3CpqZ8
                    WZgZGTPKnz2FwLY6wO7BJhJ30hGzzk6T8hhJMQN/AB342ZKnOgpRg/1glLg5A6HN
                    gwIDAQAB
                    -----END PUBLIC KEY-----
              issuer: "solo.io"
              tokenSource:
                headers:
                - header: authorization
                  prefix: "Bearer "
              claimsToHeaders:
              - claim: name
                header: x-solo-end-user
              audiences:
              - frontend
```

Requests to `$(glooctl proxy url)/l` which do not provide an authorization header are forwarded to the upstream without being authenticated. 

Current working theory is that changes to the way that envoy handles redirects are responsible for the bug

**Comment by nrjpoddar** [comment]
@ben-taussig-solo I would like this closed early next week so let's close out the PR and communicate with @sam-heilbron if we are going to slip.

**Comment by ben-taussig-solo** [comment]
Sure! I have a draft PR up [here](https://github.com/solo-io/gloo/pull/6220) in open-source Gloo which along with some enterprise changes should resolve the issue. I expect to have the issue resolved by early next week and will update the issue with my progress.

**Comment by ben-taussig-solo** [comment]
Update: The PRs linked above are fairly lightweight and sufficient to resolve the issue. There is discussion around the possibility of introducing a heftier changeset which would more significantly effect the behavior of Gloo Edge, and in turn will take longer to validate. Essentially, those changes would process route-level configuration for /all/ route plugins, as opposed to just for plugins which opt-in as proposed in the PRs above. I have opened a draft PR [here](https://github.com/solo-io/gloo/pull/6233) to test that set of changes, and have opened a dialogue with the field to discuss. Depending on the results of that conversation and how the draft PR stands up to CI, we will either:
 - Proceed with merging the original PRs (linked above), which should not take more than 1 work day
 - Proceed with the more widely reaching changeset, which could take a longer amount of time (rough estimate of ~3-5 days, depending on what edge cases we end up running into.)

**Comment by nrjpoddar** [comment]
Thanks @ben-taussig-solo! If we are going down the second route we might have to put this work on hold in favor of other high priority items so let me know how the conversations go.

**Comment by ben-taussig-solo** [comment]
Update: We are proceeding with the second approach detailed above. The corresponding PR (see [here](https://github.com/solo-io/gloo/pull/6233)) is passing CI and contains a series of tests which validate the more extensive behavioral change. Tomorrow I will get new reviews on the PR and (hopefully) merge the changes. If all goes quickly, we may be able to close this issue by EOD tomorrow. Realistically, it may take until Friday in order to finish backporting this functionality to the necessary versions.

**Comment by ben-taussig-solo** [comment]
Update: The fix has been merged to v1.12.x. Reopening the issue while I backport to v1.11.x, v1.10.x, and v1.9.x
